drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,432 @@
+			cfg->xPadConf->ucFunc = PAD38_FUNC_SEL_GPIO_6;
+			break;
+		case GPIO_7:
+			cfg->xPadConf->ucPin = PAD_45;
+			cfg->xPadConf->ucFunc = PAD45_FUNC_SEL_GPIO_7;
+			break;
+		default:
+			return -EINVAL;
+		}
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * NOTE:
+ * Pin number defined in the device tree should be actually a GPIO number
+ * to avoid overflow in high level GPIO API.
+ * `gpio_num_extended` in range of 0 to 7 matches to `port_num == 0` PADs from
+ * `find_pad` function and `gpio_num_extended` in range of 8 to 15
+ * are associated with `port_num == 1` consequently.
+ */
+static int gpio_eos_s3_configure(const struct device *dev,
+				 gpio_pin_t gpio_num_extended,
+				 gpio_flags_t flags)
+{
+	ARG_UNUSED(dev);
+	int ret;
+	GPIOCfgTypeDef gpio_cfg;
+	PadConfig pad_conf;
+	int port_num = 0;
+
+	gpio_cfg.xPadConf = &pad_conf;
+
+	/* Check for an invalid gpio number */
+	if (gpio_num_extended >= MAX_GPIOS) {
+		return -EINVAL;
+	}
+
+	/* Determine "port" number */
+	if ((int)(gpio_num_extended / MAX_GPIOS_PER_PORT) == 1) {
+		port_num = 1;
+	}
+
+	/* Get actual GPIO number */
+	uint8_t gpio_num = gpio_num_extended % 8;
+
+	/* Check for an invalid pin configuration */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	ret = find_pad(&gpio_cfg, port_num);
+
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure PAD */
+	if (flags & GPIO_PULL_UP) {
+		gpio_cfg.xPadConf->ucPull = PAD_PULLUP;
+	} else if (flags & GPIO_PULL_DOWN) {
+		gpio_cfg.xPadConf->ucPull = PAD_PULLDOWN;
+	} else {
+		gpio_cfg.xPadConf->ucPull = PAD_NOPULL;
+	}
+
+	if ((flags & GPIO_DIR_MASK) == GPIO_INPUT) {
+		gpio_cfg.xPadConf->ucMode = PAD_MODE_INPUT_EN;
+		gpio_cfg.xPadConf->ucSmtTrg = PAD_SMT_TRIG_EN;
+		HAL_PAD_Config(gpio_cfg.xPadConf);
+	} else {
+		gpio_cfg.xPadConf->ucMode = PAD_MODE_OUTPUT_EN;
+		HAL_PAD_Config(gpio_cfg.xPadConf);
+	}
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_get_raw(const struct device *dev,
+				    uint32_t *value)
+{
+	ARG_UNUSED(dev);
+	uint8_t pin_value = 0;
+	*value = 0;
+
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
+		HAL_GPIO_Read(gpion, &pin_value);
+		*value |= (pin_value == 1) ? 1 << gpion : 0;
+	}
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_set_masked_raw(const struct device *dev,
+					   uint32_t mask,
+					   uint32_t value)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = ((output_states & ~mask) | (value & mask));
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
+		if (target_value & 1) {
+			HAL_GPIO_Write(gpion, 1);
+		} else {
+			HAL_GPIO_Write(gpion, 0);
+		}
+		target_value = target_value >> 1;
+	}
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_set_bits_raw(const struct device *dev,
+					 uint32_t value)
+{
+	ARG_UNUSED(dev);
+
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
```

Comment:
I'm a bit confused by this.  You only have one device, but above it sounded like you have two ports with eight pins each, which are identified by port*8 + pin.  But this only iterates over 8 pins.

The pin configure code seems to respect the extended GPIO num to select a pad configuration given pin 7 or pin 15, but it seems like this would not do anything for pins on port 1.

I guess it must work....

Reply:
Let me explain. There are total 16 physical pads/pins which can be configured as GPIOs. EOS S3 has only 8 GPIOs, each GPIO has a corresponding pair of physical pins and only one pin can be picked from a pair and used as GPIO. So it is possible to configure only 8 GPIOs connected to 8 pins. Controller needs only information about GPIO number (0-7) which should be read or written, beacuse the corresponding pin is assigned on the configure step.